File: tasks/translation/output/Qwen/QwQ-32B-AWQ/codenet/direct/Java/C++/s749703010_commented.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

vector<vector<vector<int>>> memo;
vector<vector<int>> a;
int n;

int dp(int flipped, int idx, int msk) {
    if (msk == (1 << n) - 1) return 0;
    if (memo[flipped][idx][msk] != -1) return memo[flipped][idx][msk];
    int x = __builtin_popcount(msk);
    int ff = 0;
    int ans = 1e9;
    for (int i = 0; i < n; i++) {
        if ((msk & (1 << i)) == 0) {
            int nf = x % 2;
            int nf_plus = (nf + (ff % 2)) % 2;
            if (idx == n || a[flipped][idx] <= a[nf_plus][i]) {
                ans = min(ans, ff + dp(nf_plus, i, msk | (1 << i)));
            }
            ff++;
        } else {
            x--;
        }
    }
    return memo[flipped][idx][msk] = ans;
}

int main() {
    ios_base::sync_with_stdio(0);
    cin.tie(0);
    cin >> n;
    a.resize(2, vector<int>(n));
    for (int i = 0; i < n; i++) cin >> a[0][i];
    for (int i = 0; i < n; i++) cin >> a[1][i];
    memo.resize(2);
    for (int i = 0; i < 2; i++) {
        memo[i].resize(n + 1);
        for (int j = 0; j <= n; j++) {
            memo[i][j].resize(1 << n, -1);
        }
    }
    int ans = dp(0, n, 0);
    cout << (ans >= 1e9 ? -1 : ans) << endl;
    return 0;
}
// End of Code

File: tasks/translation/output/Qwen/QwQ-32B-AWQ/codenet/direct/Java/Go/s749703010_commented.go
package main

import (
	"bufio"
	"fmt"
	"math/bits"
	"os"
	"strconv"
	"strings"
)

func main() {
	reader := bufio.NewReader(os.Stdin)
	n, _ := strconv.Atoi(readLine(reader))
	a := make([][]int, 2)
	a[0] = make([]int, n)
	a[1] = make([]int, n)

	nums0, _ := readInts(reader)
	for i := 0; i < n; i++ {
		a[0][i] = nums0[i]
	}

	nums1, _ := readInts(reader)
	for i := 0; i < n; i++ {
		a[1][i] = nums1[i]
	}

	memo := make([][][]int, 2)
	for i := range memo {
		memo[i] = make([][]int, n+1)
		for j := range memo[i] {
			memo[i][j] = make([]int, 1<<n)
			for k := range memo[i][j] {
				memo[i][j][k] = -1
			}
		}
	}

	var dp func(flip, idx, msk int) int
	dp = func(flip, idx, msk int) int {
		if msk == (1<<n)-1 {
			return 0
		}
		if memo[flip][idx][msk] != -1 {
			return memo[flip][idx][msk]
		}
		x := bits.OnesCount(uint(msk))
		ans := 1000000000
		ff := 0
		for i := 0; i < n; i++ {
			if (msk&(1<<i)) == 0 {
				nf := x % 2
				nextFlip := (nf + (ff%2)) % 2
				nextAVal := a[nextFlip][i]
				var conditionMet bool
				if idx == n {
					conditionMet = true
				} else {
					currentA := a[flip][idx]
					conditionMet = currentA <= nextAVal
				}
				if conditionMet {
					res := ff + dp(nextFlip, i, msk|(1<<i))
					if res < ans {
						ans = res
					}
				}
				ff++
			} else {
				x--
			}
		}
		memo[flip][idx][msk] = ans
		return ans
	}

	ans := dp(0, n, 0)
	if ans >= 1000000000 {
		fmt.Println(-1)
	} else {
		fmt.Println(ans)
	}
}

func readLine(r *bufio.Reader) string {
	line, _, _ := r.ReadLine()
	return string(line)
}

func readInts(r *bufio.Reader) ([]int, error) {
	line, _, err := r.ReadLine()
	if err != nil {
		return nil, err
	}
	parts := strings.Fields(string(line))
	nums := make([]int, len(parts))
	for i, s := range parts {
		num, _ := strconv.Atoi(s)
		nums[i] = num
	}
	return nums, nil
}

// End of Code
